swarm-derive/src/lib.rs

```diff
@@ -896,7 +896,11 @@ fn parse_attributes(ast: &DeriveInput) -> Result<BehaviourAttributes, TokenStrea
                 continue;
             }
 
-            if meta.path().is_ident("out_event") {
+            if meta.path().is_ident("to_swarm") || meta.path().is_ident("out_event") {
+                if meta.path().is_ident("out_event") {
+                    log::warn!("The 'out_event' key is deprecated, use 'to_swarm' instead",);
```

Comment:
Where is this warning going to show up? `log` is only an API crate and needs a concrete implementation like `env_logger` to actually print the messages. Does cargo automatically configure a logger when it runs proc-macros?

Reply:
@thomaseizinger I thought so, but you're right it's not, I added `env_logger::init();` to put it simply, in this way the warning should be displayed correctly.

Reply:
> should be displayed correctly.

@tcoratger Have you confirmed that it **is** actually displayed? :)

I am asking because there is an entire [nightly library feature](https://github.com/rust-lang/rust/issues/54140) that deals with diagnostics in proc macros and I think it is not as easy as just emitting a `log` :)

I think we can work around this by emitting code that has deprecation warnings on it. Here is an idea:

1. If the user users `out_event`, also generate a trait: `trait OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm { }`
2. Also put a `#[deprecated]` attribute on this trait
3. Implement the trait for the generated out event

I _think_ this should then show up as a deprecation warning in the user's code.

Reply:
@thomaseizinger Yes, then in fact the `println!` was indeed working badly, I had a warning but only when compiling so it was wrong. I implemented a trait for this and it should do the trick:

```
warning: use of deprecated trait `OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm`: The 'out_event' key is deprecated, use 'to_swarm' instead.
   --> swarm-derive/src/lib.rs:907:26
    |
907 |                     impl OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm for Meta {}
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(deprecated)]` on by default

warning: `libp2p-swarm-derive` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.23s
     Running `target/debug/distributed-key-value-store`
Listening in "/ip4/127.0.0.1/tcp/63573"
Listening in "/ip4/192.168.25.22/tcp/63573"
```

Anyway I guess it's ok if it's not perfect because the deprecation will soon be totally removed right?

Reply:
@thomaseizinger Can you enlighten me on the subject, I am not very familiar with proc macros, I noticed the following things:
- When I do a simple `println!` I thought it was the right approach but in fact the warning is generated only at the first compilation but not at each use of the code, because I imagine that it is a piece of code that runs only at compilation.
- On the another hand when I use an approach like:
```rust
if meta.path().is_ident("out_event") {
  #[deprecated(
      since = "0.33.0",
      note = "The 'out_event' key is deprecated, use 'to_swarm' instead."
  )]
  trait OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm {}

  impl OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm for Meta {}
}
```
then the deprecation warning appears each time even if we use `to_swarm`, this is normal because we face a general trait implementation that is deprecated and this is not what we want. Any advice here?

Reply:
If you wanted to write a test for this, I'd recommend looking at `trybuild`: https://github.com/dtolnay/trybuild

Reply:
@thomaseizinger Then I propose something like [this commit](https://github.com/libp2p/rust-libp2p/pull/3848/commits/5d3ac92b46862b3d34860faa25e1976463963c2b):
- Creating a `is_out_event` boolean to check during the function if `out_event` is found
- If so, then trying to wrap a trait and its implementation inside a quote for the output type, given thanks to `attributes` with the following piece of code:
```rust
if is_out_event {
    let out_event_type: syn::Type = match attributes.user_specified_out_event {
        Some(ref path) => {
            let path_str = quote! { #path }.to_string();

            syn::parse_str(&path_str).unwrap()
        }
        None => syn::parse_str("()").unwrap(),
    };

    quote! {
        #[deprecated(
            since = "0.33.0",
            note = "The 'out_event' key is deprecated, use 'to_swarm' instead."
        )]
        trait OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm{};

        impl OutEventAttributeIsDeprecatedAndHasBeenRenamedToToSwarm for #out_event_type {}
    };
}
```

- I didn't create any particular test since we have the examples for that :) For instance, in the `distributed-key-value-store` folder, we can find examples for this situation:
```rust
// We create a custom network behaviour that combines Kademlia and mDNS.
#[derive(NetworkBehaviour)]
#[behaviour(out_event = "MyBehaviourEvent")]
struct MyBehaviour {
    kademlia: Kademlia<MemoryStore>,
    mdns: mdns::async_io::Behaviour,
}
```
In this case, I left `out_event` but I don't see a warning appear when I run the example, despite the fact that the code goes into my `is_out_event` condition, do you think this is the right direction?

Reply:
I'll have a look.

Reply:
I couldn't get this to work. For some reason, I am not seeing _any_ warnings that should be produced for the generated code. I swear I saw some of them in the past (which is why we have `allow(clippy)` attributes in there).

I've found https://github.com/ggwpez/proc-macro-warning but couldn't get it to work. Let's just leave it out and document that it is deprecated.

Reply:
@thomaseizinger Yes on my side also I looked a lot and I did not find anything conclusive. You mean doing an implementation like:
```rust
if meta.path().is_ident("to_swarm") || meta.path().is_ident("out_event") {
  if meta.path().is_ident("out_event") {
      quote! {
          macro_rules! emit_warning {
              ($span:expr, $($tts:tt)*) => {
                  println!("warning: {}", format_args!($($tts)*));
              };
          }
      };

      let emit_warning_call = quote! {
          emit_warning!(#meta, "The 'out_event' key is deprecated, use 'to_swarm' instead.");
      };

      use quote::ToTokens;
      quote! {
          #emit_warning_call
      }
      .to_token_stream();
  }
....
}
```

I've tried that a little bit yesterday switching my rustup to nightly and did work properly, what do you think?

Reply:
Ha, that is also a creative idea but not quite what I meant :)

We can just depend on the `proc-macro-error` crate directly and use their `emit_warning!` macro! On nightly, that should allow us to issue a diagnostic, similar to how the Rust compiler or clippy issues them.

Reply:
It seems that we can have warnings on Rust stable too!

See https://github.com/ggwpez/proc-macro-warning/issues/2#issuecomment-1535390669.